### PR TITLE
Make sure attributes used on subsequent shader stages are initialized

### LINKED ISSUE #1895: Hatsune Miku Mega39 - AMD GPU Incorrect/Broken Render
### What's the issue you encountered?

Loading in any module will produce eyes that aren't rendered correctly on an AMD GPU.</br>

Incorrect Render 1</br>

Another notably example is one particular module that has white eyes, though it is rendered with the same error shown above</br>
Incorrect Render 2</br>

There is a seperate issue opened currently on yuzu that has a similar error like the one shown that I can link if asked for.

### How can the issue be reproduced?

Either enter the customization menu or use the default module and enter a song (In PV mode or rhythm mode)

### Log file

[Ryujinx_1.0.6291_2021-01-11_12-50-56.log](https://github.com/Ryujinx/Ryujinx/files/5797792/Ryujinx_1.0.6291_2021-01-11_12-50-56.log)

### Environment?

- Ryujinx version: 1.0.6291</br>
- Game version: 1.0.0</br>
- System Specs:
  - OS: Windows 7
  - CPU: i3-6100
  - GPU: AMD R7 360 2GB VRAM
  - RAM: 8GB
- Applied Mods : No

### Additional context?

Additional info about your environment:</br>
None that I'm aware

### PR DESCRIPTION
**Pull request by gdkchan**
Sometimes, an shader attribute is consumed on a later stage, but the previous stage does not write it. On NVIDIA, all attributes that are not written to are initialized with a default value of (0, 0, 0, 1). This is not the case for other vendors however, and according to the spec, the values on them are "undefined". This causes problems for AMD and Intel as the attributes have different values.

The current approach taken is assuming all the first 16 attributes are used, so initializes the first 16.  That's because on desktop, usually only 16 attributes are supported (at least on OpenGL, Vulkan might allow more?). This causes shaders to fail to compile if the previous stage does not write the attribute at all. For example, one shader in Yo-kai Watch 4 fails to compile due to this, I'm not sure for what the shader is used though, the game seems to render fine regardless.

Additionally, there is another problem where partially written attributes are not initialized. So if the vertex shader writes X, Y and Z, and the  fragment shader reads W, the value will be undefined. This causes eyes to be solid blue on Hatsune Miku, and shadows to break in Zelda Link's Awakening on Intel and AMD.

This change fixes all the issues above by propagating attribute use information backwards (from later stages to early ones, like from fragment to vertex etc). That way, all attributes that are used on the subsequent stages can be properly initialized. Attribute use information is now computed very early (on the decode step) rather than later (on the structured program construction step), this was done to allow the uses to be propagated after decode, but before translation. Attribute initialization is also done earlier now, it is done at the IR level instead of the backend, this reduces code duplication between GLSL and SPIR-V.

So, to sum it up, those issues should be fixed:
- Shadows not working on Zelda Link's awakening (Intel and AMD).
- Hatsune Miku eyes being black (Intel and AMD).
- Shader failing to compile on Yo-kai Watch 4 (all vendors, no visible effect though?).

Probably more...

Would be nice to get a confirmation that those games were fixed on AMD, as I did not test them *on AMD*, and they still don't render properly on Intel OpenGL due to other issues (I will be fixing some of them soon), also helps Vulkan.

Fixes #1895 and #1109.

### CODE REVIEW

**Comment by SimpleName140** [comment]
Tested this PR, and can confirm on my hardware that it does fix it. Thanks for the fix :+1: 
AmdFix1
AmdFix2

**Review by Thog — APPROVED** [review]

**Review by AcK77 — APPROVED** [review]

**Comment by RinMaru** [comment]
this works on AMD i can confirm. RX 480
